File: projects/rxtwo/src/main/java/com/azunitech/rx/rxtwo/usecases/CreateProductImpl.java
package com.azunitech.rx.rxtwo.usecases;

import com.azunitech.rx.rxtwo.appliation.IDGenerator;
import com.azunitech.rx.rxtwo.domain.Product;
import com.azunitech.rx.rxtwo.domain.ProductGateway;
import com.azunitech.rx.rxtwo.usecases.apis.CreateProduct;
import org.springframework.stereotype.Component;

@Component
public class CreateProductImpl implements CreateProduct {
  final ProductGateway productGateway;

  private final IDGenerator idGenerator;

  public CreateProductImpl(ProductGateway productGateway, IDGenerator idGenerator) {
    this.productGateway = productGateway;
    this.idGenerator = idGenerator;
  }

  @Override
  public Response create(Request request) {
    String id = idGenerator.createID();
    Product product = Product.builder().id(id).name(request.getName()).build();
    Product createdProduct = productGateway.createProduct(product);
    return Response.builder().product(createdProduct).build();
  }
}

File: projects/cleancode/src/main/kotlin/com/azunitech/cleancode/usercases/CreateProductImpl.kt
package com.azunitech.cleancode.usercases

import com.azunitech.cleancode.domain.Product
import com.azunitech.cleancode.domain.ProductGateway
import com.azunitech.cleancode.usercases.apis.CreateProduct
import com.azunitech.cleancode.usercases.apis.NotifyEvent
import org.springframework.stereotype.Component
import java.util.*

@Component
class CreateProductImpl(val productGateway: ProductGateway?, val raiser: NotifyEvent) : CreateProduct {
    override fun create(request: CreateProduct.Request): CreateProduct.Response {
        val id = productGateway!!.create(request.toDomain())
        raiser.notice(NotifyEvent.ProductCreatedEvent(id))
        return CreateProduct.Response(id)
    }
}

fun CreateProduct.Request.toDomain(): Product {
    return Product(UUID.randomUUID().toString(), name)
}